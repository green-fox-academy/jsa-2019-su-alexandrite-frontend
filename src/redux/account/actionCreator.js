import { LOCAL_URL } from 'react-native-dotenv';
import {
  LOGIN_USER_FAILURE,
  LOGIN_USER_SUCCESS,
  LOGOUT_SUCCESS,
} from './actionType';

const loginUserSuccess = (payload) => ({
  type: LOGIN_USER_SUCCESS,
  payload,
});

export const loginUserFail = (payload) => ({
  type: LOGIN_USER_FAILURE,
  payload,
});

const loginUser = (username, password) => (dispatch) => {
  const loginURL = `${LOCAL_URL}/users/login`;
  fetch(loginURL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      username,
      password,
    }),
  })
    .then((response) => {
      if (response.status === 200) {
        return response.json();
      }
      throw new Error('Oops! there is something wrong with your app');
    })
    .then((response) => {
      dispatch(loginUserSuccess(response));
    })
    .catch((error) => dispatch(loginUserFail(error)));
};

export const logOut = () => ({
  type: LOGOUT_SUCCESS,
});

export default loginUser;
