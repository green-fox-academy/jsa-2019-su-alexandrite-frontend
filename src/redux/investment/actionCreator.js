import { SERVER_URL, API_URL, API_KEY } from 'react-native-dotenv';
import {
  FETCH_PORTFOLIO_DETAILS_START,
  FETCH_PORTFOLIO_DETAILS_FAIL,
  FETCH_PORTFOLIO_DETAILS_SUCCESS,
} from './actionType';
import { round, moneyAmount2String } from '../../common/numbers';

export const fetchPortfolioDetailsStart = () => ({
  type: FETCH_PORTFOLIO_DETAILS_START,
});

export const fetchPortfolioDetailsFail = (payload) => ({
  type: FETCH_PORTFOLIO_DETAILS_FAIL,
  payload,
});

export const fetchPortfolioDetailsSuccess = (payload) => ({
  type: FETCH_PORTFOLIO_DETAILS_SUCCESS,
  payload,
});

export const calculatePortfolioValue = () => (dispatch, getState) => {
  const serverUrl = new URL(`${SERVER_URL}/investments/user/`);
  const headers = new Headers();
  const { accessToken } = getState().user;
  headers.append('authorization', `Bearer ${accessToken}`);
  dispatch(fetchPortfolioDetailsStart());
  fetch(serverUrl, { headers })
    .then((response) => {
      if (!response.ok) {
        switch (response.status) {
          case 401:
            throw new Error('Sorry, we cannot validate your identity. Please login and try again.');
          default:
            throw new Error('Oops, there\'s something wrong with our app.');
        }
      }
      return response.json();
    })
    .then((stocks) => {
      const symbols = stocks.map((stock) => stock.symbol);
      const apiUrl = new URL(`${API_URL}/stock/market/batch`);
      apiUrl.searchParams.append('symbols', symbols);
      apiUrl.searchParams.append('types', 'price,company,logo');
      apiUrl.searchParams.append('token', API_KEY);
      fetch(apiUrl)
        .then((res) => {
          if (!res.ok) {
            throw new Error('Oops, there\'s something wrong with our app.');
          }
          return res.json();
        })
        .then((res) => {
          const instruments = {};
          let totalValue = stocks
            .map(({ shares, symbol }) => {
              instruments[symbol] = {
                logo: res[symbol].logo.url,
                company: res[symbol].company.companyName,
                description: res[symbol].company.description,
                marketValue: round(shares * res[symbol].price),
              };
              return shares * res[symbol].price;
            })
            .reduce((a, b) => a + b);
          totalValue = moneyAmount2String((totalValue));
          dispatch(fetchPortfolioDetailsSuccess({
            totalValue,
            stocks,
            instruments,
          }));
        });
    })
    .catch((err) => dispatch(fetchPortfolioDetailsFail(err)));
};
