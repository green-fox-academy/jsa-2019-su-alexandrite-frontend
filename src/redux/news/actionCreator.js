import {
  SERVER_URL,
} from 'react-native-dotenv';
import {
  FETCH_INVESTED_STOCKS_START,
  FETCH_INVESTED_STOCKS_FAIL,
  FETCH_INVESTED_STOCKS_SUCCESS,
} from './actionType';

export const fetchInvestedStocksStart = () => ({
  type: FETCH_INVESTED_STOCKS_START,
});

export const fetchInvestedStocksFail = (payload) => ({
  type: FETCH_INVESTED_STOCKS_FAIL,
  payload,
});

export const fetchInvestedStocksSuccess = (payload) => ({
  type: FETCH_INVESTED_STOCKS_SUCCESS,
  payload,
});

export const fetchInvestedStocks = () => async (dispatch, getState) => {
  try {
    const serverUrl = new URL(`${SERVER_URL}/investments/user/`);

    const headers = new Headers();
    const { accessToken } = getState().user;
    headers.append('authorization', `Bearer ${accessToken}`);

    dispatch(fetchInvestedStocksStart());
    const investmentsResult = await fetch(serverUrl, { headers });
    if (!investmentsResult.ok) {
      switch (investmentsResult.status) {
        case 401:
          throw new Error('Sorry, we cannot validate your identity. Please login and try again.');
        default:
          throw new Error('Oops, there\'s something wrong with our app.');
      }
    }

    const stocks = await investmentsResult.json();
    const symbols = stocks.map(({ symbol }) => symbol);
    dispatch(fetchInvestedStocksSuccess(
      symbols,
    ));
  } catch (err) {
    dispatch(fetchInvestedStocksFail(err));
  }
};
