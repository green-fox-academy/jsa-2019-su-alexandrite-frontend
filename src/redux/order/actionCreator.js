import { SERVER_URL } from 'react-native-dotenv';

import {
  PURCHASE_STOCK_SUCCESS,
  PURCHASE_STOCK_FAIL,
} from './actionType';

const purchaseStockSuccess = (payload) => ({
  type: PURCHASE_STOCK_SUCCESS,
  payload,
});

const purchaseStockFail = (payload) => ({
  type: PURCHASE_STOCK_FAIL,
  error: payload,
});

const purchaseStock = (symbol, shares, type, accessToken) => (dispatch) => {
  const orderUrl = new URL(`${SERVER_URL}/order`);
  fetch(orderUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify(
      {
        symbol,
        shares,
        type,
      },
    ),
  })
    .then((response) => {
      if (response.status !== 201) {
        switch (response.status) {
          case 401:
            throw new Error('Sorry, we cannot validate your identity. Please login and try again.');
          default:
            throw new Error('Oops! there\'s something wrong with our app.');
        }
      }
      return response.json();
    })
    .then(() => dispatch(purchaseStockSuccess()))
    .catch((error) => dispatch(purchaseStockFail(error.message)));
};

export default purchaseStock;
