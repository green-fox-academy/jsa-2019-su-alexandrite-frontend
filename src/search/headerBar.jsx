import React, { useState, useEffect } from 'react';
import { SafeAreaView, TextInput } from 'react-native';
import { useDispatch } from 'react-redux';
import { fetchData } from '../redux/search/actionCreater';

export default function SearchBar() {
  const [value, onChangeText] = useState('');
  const dispatch = useDispatch();
  useEffect(() => {
    let timer;
    if (value.length >= 2) {
      timer = setTimeout(() => {
        dispatch(fetchData(value));
        alert('times up');
      }, 3000);
    }
    return () => {
      if (timer) clearTimeout(timer);
    };
  }, [value]);

  function handleChange(text) {
    onChangeText(text);
    // clearTimeout(timer);
    // if (value.length >= 2) {
    //   timer = setTimeout(() => {
    //     dispatch(fetchData(value));
    //     alert('times up');
    //   }, 2000);
    // }
  }

  return (
    <SafeAreaView>
      <TextInput
        style={{ height: 40, backgroundColor: '#566ed3' }}
        onChangeText={(text) => handleChange(text)}
        value={value}
      />
    </SafeAreaView>
  );
}
