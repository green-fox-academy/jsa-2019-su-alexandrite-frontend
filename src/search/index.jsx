import React from 'react';
import { View, Text } from 'react-native';
import SearchBar from './headerBar';

const navigationOptions = {
  header: () => <SearchBar />,
};

const search = () => (
  <View>
    <Text>Search Screen</Text>
  </View>
);

search.navigationOptions = navigationOptions;

export default search;
