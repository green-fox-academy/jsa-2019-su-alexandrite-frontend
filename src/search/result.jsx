import React from 'react';
import { useSelector } from 'react-redux';
import { View, Text, Image } from 'react-native';
import { TouchableHighlight } from 'react-native-gesture-handler';
import styles from './styles';

import tradeIcon from '../../assets/icons/search-page/trade.png';
import detailIcon from '../../assets/icons/search-page/detail.png';

export default () => {
  const isLoading = useSelector((state) => state.search.isLoading);
  const searchResult = useSelector((state) => state.search.result);
  const error = useSelector((state) => state.search.error);

  if (isLoading) {
    return <Text>Loading</Text>;
  }
  if (error) {
    return <Text>{error.message}</Text>;
  }
  return (
    <View>
      {searchResult.map((info) => (
        <View key={info.symbol} style={styles.resultList}>
          <View>
            <Text style={styles.stockName}>{info.symbol}</Text>
            <Text style={styles.stockExchange}>{info.exchange}</Text>
          </View>
          <View style={styles.resultButton}>
            <TouchableHighlight>
              <Image source={tradeIcon} />
            </TouchableHighlight>
            <TouchableHighlight style={styles.detailButton}>
              <Image source={detailIcon} />
            </TouchableHighlight>
          </View>
        </View>
      ))}
    </View>
  );
};
