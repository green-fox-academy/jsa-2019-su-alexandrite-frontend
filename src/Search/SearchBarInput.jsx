import React, { useEffect } from 'react';
import {
  TextInput,
} from 'react-native';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import styles from './styles';
import { searchStockData } from '../redux/search/actionCreator';

const SearchBarInput = ({
  text,
  setText,
}) => {
  const dispatch = useDispatch();
  useEffect(() => {
    let timer;
    if (text.length >= 2) {
      timer = setTimeout(() => {
        dispatch(searchStockData(text));
      }, 300);
    }
    return () => {
      if (timer) clearTimeout(timer);
    };
  }, [text]);

  return (
    <TextInput
      testID="search-text-input"
      style={styles.searchBar}
      clearButtonMode="while-editing"
      onChangeText={setText}
      value={text}
    />
  );
};


SearchBarInput.propTypes = {
  text: PropTypes.string.isRequired,
  setText: PropTypes.func.isRequired,
};


export default SearchBarInput;
