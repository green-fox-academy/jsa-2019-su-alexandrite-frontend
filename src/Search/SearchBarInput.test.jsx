import React from 'react';
import { useDispatch } from 'react-redux';
import { shallow } from 'enzyme';
import renderer, { act } from 'react-test-renderer';
import { searchStockData } from '../redux/search/actionCreator';
import SearchBarInput from './SearchBarInput';

jest.mock('react-redux');
jest.mock('../redux/search/actionCreator');

describe('<SearchBarInput />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const wrapper = shallow(<SearchBarInput text="" setText={jest.fn()} />);
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const tree = renderer.create(
      <SearchBarInput text="" setText={jest.fn()} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Should not trigger search if query length is le 3', () => {
    jest.useFakeTimers();
    const mockDispatch = jest.fn(() => { });
    useDispatch.mockReturnValue(mockDispatch);
    searchStockData.mockReturnValueOnce('test_action');
    const wrapper = renderer.create(
      <SearchBarInput text="" setText={jest.fn()} />,
    );
    act(() => {
      wrapper.update(<SearchBarInput text="1" setText={jest.fn()} />);
    });
    expect(setTimeout).not.toHaveBeenCalled();
    renderer.act(() => {
      jest.advanceTimersByTime(300);
    });
    expect(mockDispatch).not.toHaveBeenCalled();
    expect(searchStockData).not.toHaveBeenCalledWith();
    renderer.act(() => {
      wrapper.unmount();
    });
  });

  it('Should trigger search if query length is gt 3', () => {
    jest.useFakeTimers();
    const mockDispatch = jest.fn(() => { });
    useDispatch.mockReturnValue(mockDispatch);
    searchStockData.mockReturnValueOnce('test_action');
    const wrapper = renderer.create(
      <SearchBarInput text="" setText={jest.fn()} />,
    );
    act(() => {
      wrapper.update(<SearchBarInput text="123" setText={jest.fn()} />);
    });
    expect(setTimeout).toHaveBeenCalledTimes(1);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 300);
    renderer.act(() => {
      jest.advanceTimersByTime(300);
    });
    expect(mockDispatch).toHaveBeenCalledWith('test_action');
    expect(searchStockData).toHaveBeenCalledWith('123');
    renderer.act(() => {
      wrapper.unmount();
    });
  });
});
