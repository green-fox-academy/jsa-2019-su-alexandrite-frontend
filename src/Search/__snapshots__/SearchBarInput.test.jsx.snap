// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBarInput /> Should render and match the snapshot 1`] = `
<TextInput
  allowFontScaling={true}
  clearButtonMode="while-editing"
  onChangeText={[MockFunction]}
  rejectResponderTermination={true}
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 50,
      "flex": 1,
      "fontSize": 12,
      "paddingHorizontal": 11,
      "paddingVertical": 8,
    }
  }
  testID="search-text-input"
  underlineColorAndroid="transparent"
  value=""
/>
`;
